core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/LayoutConstraintEvaluator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.operator.scalar.TryFunction;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ConstraintEvaluator;
+import io.trino.spi.predicate.NullableValue;
+import io.trino.sql.planner.ExpressionInterpreter;
+import io.trino.sql.planner.LookupSymbolResolver;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SymbolsExtractor;
+import io.trino.sql.planner.TypeAnalyzer;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.NullLiteral;
+
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.intersection;
+
+public class LayoutConstraintEvaluator
+        implements ConstraintEvaluator
+{
+    private final Map<Symbol, ColumnHandle> assignments;
+    private final ExpressionInterpreter evaluator;
+    private final Set<ColumnHandle> arguments;
+
+    public LayoutConstraintEvaluator(Metadata metadata, TypeAnalyzer typeAnalyzer, Session session, TypeProvider types, Map<Symbol, ColumnHandle> assignments, Expression expression)
+    {
+        this.assignments = assignments;
```

Comment:
`rnn` and `ImmutableMap.copyOf`  (now that the class is public, it's more important)